rust/src/nasl/builtin/cert/mod.rs

```diff
@@ -0,0 +1,415 @@
+    /// for further operations.
+    #[nasl_function]
+    fn cert_open(&self, cert: &[u8]) -> Result<usize, FunctionErrorKind> {
+        if let Ok(cert) = X509Certificate::from_der(cert) {
+            return Ok(self.insert(cert));
+        }
+        if let Ok(cert) = X509Certificate::from_pem(cert) {
+            return Ok(self.insert(cert));
+        }
+        if let Ok(cert) = X509Certificate::from_ber(cert) {
+            return Ok(self.insert(cert));
+        }
+
+        Err(FunctionErrorKind::WrongArgument(
+            "The given string is not a valid DER, BER or PEM encoded X.509 certificate."
+                .to_string(),
+        ))
+    }
+
+    /// Release a certificate object.
+    ///
+    /// Takes a cert identifier as returned by cert_open and releases the
+    /// associated resources.
+    #[nasl_function]
+    fn cert_close(&self, cert_handle: usize) {
+        let mut handle = self.0.write().unwrap();
+        handle.certs.remove(&cert_handle);
+    }
+
+    fn subject(cert: &X509Certificate, idx: usize) -> Option<String> {
+        // The error originates from the io::Write trait. Internally a Vec is used, which
+        // implementation of that trait is infallible. Therefore we can unwrap here.
+        let der = cert.encode_der().unwrap();
+        let (_, cert) = x509_parser::parse_x509_certificate(&der).unwrap();
+
+        if idx == 0 {
+            Some(cert.subject.to_string())
+        } else {
+            cert.subject_alternative_name()
+                .ok()
+                .flatten()
+                .and_then(|san| san.value.general_names.get(idx - 1))
+                .map(|san| Some(san.to_string()))
+                .unwrap_or(None)
+        }
+    }
+
+    fn issuer(cert: &X509Certificate, idx: usize) -> Option<String> {
+        let subject = cert.issuer_name();
+        subject.get(idx).map(|entry| {
+            entry
+                .iter()
+                .filter_map(|val| {
+                    val.value.to_string().ok().map(|value| {
+                        format!("{}={}", subject_oid_to_name(&val.typ.to_string()), value)
+                    })
+                })
+                .collect::<Vec<_>>()
+                .join(", ")
+        })
+    }
+
+    fn hostnames(cert: &X509Certificate) -> Vec<String> {
+        let mut ret = vec![];
+        if let Some(cn) = cert.subject_common_name() {
+            ret.push(cn);
+        }
+
+        let der = cert.encode_der().unwrap();
+        let (_, cert) = x509_parser::parse_x509_certificate(&der).unwrap();
+
+        if let Ok(Some(san)) = cert.subject_alternative_name() {
+            for name in san.value.general_names.iter() {
+                if let GeneralName::DNSName(dns) = name {
+                    ret.push(dns.to_string());
+                }
+            }
+        }
+
+        ret
+    }
+
+    fn key_size(cert: &X509Certificate) -> Option<i64> {
+        if let Some(algorithm) = cert.key_algorithm() {
```

Comment:
nit: I think it would be easier to read when it would be written with `?` operator:
```rust
let algorithm = cert.key_algorithm()?;
match algorithm {
 ...
}
```